okhttp/src/main/java/okhttp3/internal/platform/Platform.kt

```diff
@@ -0,0 +1,305 @@
+  open fun logCloseableLeak(message: String, stackTrace: Any?) {
+    var logMessage = message
+    if (stackTrace == null) {
+      logMessage += " To see where this was allocated, set the OkHttpClient logger level to FINE: " + "Logger.getLogger(OkHttpClient.class.getName()).setLevel(Level.FINE);"
+    }
+    log(WARN, logMessage, stackTrace as Throwable?)
+  }
+
+  open fun buildCertificateChainCleaner(trustManager: X509TrustManager): CertificateChainCleaner {
+    return BasicCertificateChainCleaner(buildTrustRootIndex(trustManager))
+  }
+
+  fun buildCertificateChainCleaner(sslSocketFactory: SSLSocketFactory): CertificateChainCleaner {
+    val trustManager = trustManager(sslSocketFactory) ?: throw IllegalStateException(
+        "Unable to extract the trust manager on "
+            + Platform.get()
+            + ", sslSocketFactory is "
+            + sslSocketFactory.javaClass)
+
+    return buildCertificateChainCleaner(trustManager)
+  }
+
+  open fun buildTrustRootIndex(trustManager: X509TrustManager): TrustRootIndex {
+    return BasicTrustRootIndex(*trustManager.acceptedIssuers)
+  }
+
+  open fun configureSslSocketFactory(socketFactory: SSLSocketFactory) {}
+
+  override fun toString(): String {
+    return javaClass.simpleName
+  }
+
+  companion object {
+    private val PLATFORM = findPlatform()
+    val INFO = 4
+    val WARN = 5
+    private val logger = Logger.getLogger(OkHttpClient::class.java.name)
+
+    fun get(): Platform {
+      return PLATFORM
+    }
+
+    fun alpnProtocolNames(protocols: List<Protocol>): List<String> {
+      val names = ArrayList<String>(protocols.size)
+      var i = 0
+      val size = protocols.size
+      while (i < size) {
+        val protocol = protocols[i]
+        if (protocol === Protocol.HTTP_1_0) {
+          i++
+          continue
+        } // No HTTP/1.0 for ALPN.
+        names.add(protocol.toString())
+        i++
+      }
+      return names
+    }
+
+    // mainly to allow tests to run cleanly
+    // check if Provider manually installed
+    val isConscryptPreferred: Boolean
+      get() {
+        if ("conscrypt" == Util.getSystemProperty("okhttp.platform", null)) {
+          return true
+        }
+        val preferredProvider = Security.getProviders()[0].name
+        return "Conscrypt" == preferredProvider
+      }
+
+    /** Attempt to match the host runtime to a capable Platform implementation.  */
+    private fun findPlatform(): Platform {
+      val android = AndroidPlatform.buildIfSupported()
+
+      if (android != null) {
+        return android
+      }
+
+      if (isConscryptPreferred) {
+        val conscrypt = ConscryptPlatform.buildIfSupported()
+
+        if (conscrypt != null) {
+          return conscrypt
+        }
+      }
+
+      val jdk9 = Jdk9Platform.buildIfSupported()
+
+      if (jdk9 != null) {
+        return jdk9
+      }
+
+      val jdkWithJettyBoot = Jdk8WithJettyBootPlatform.buildIfSupported()
+
+      return jdkWithJettyBoot ?: Platform()
```

Comment:
probably want to revert this to an `if (jdkWithJettyBoot != null)` so that the comment below can go on a `return Platform()`